### PR TITLE
Fix #5818: Add comprehensive tests for EditTextInputAction append/rep…

### LINKED ISSUE #5818: [BUG]: Fix Robolectric append/replaceText handling in EditTextInputAction and add corresponding tests.
### Describe the bug

This issue tracks the bug fix and upcoming tests related to `EditTextInputAction` used in test infrastructure for EditText input.

### 1. The Issue

- [x]  There was a bug in the `EditTextInputAction` class where appending or replacing text didn’t behave correctly in Robolectric environments. Specifically: (addressed in [PR #5617 ](https://github.com/oppia/oppia-android/pull/5617)) 
- `appendText` does not work correctly as it replaces the text rather than appending it.

- Both actions incorrectly called `EditText.setText(text)`, meaning that even when `appendText()` was used, it replaced the entire content instead of appending to it.
- This issue was masked by passing transitive test cases, but broke the intended behavior of append vs. replace differentiation.
- The logic for handling `appendText()` and `replaceText()` needed to differentiate between Robolectric and Espresso environments.

---

### 2. Adding Corresponding Tests

- [ ] Tests still need to be added to ensure the behavior of `EditTextInputAction` remains correct across test environments.

To achieve this, we should:
- Add a dedicated test class (similar to the one for [TextInputAction](https://github.com/oppia/oppia-android/blob/d25fe1f0f29688f5ed485503bc0a9c4bfff2d73f/testing/src/main/java/org/oppia/android/testing/espresso/TextInputAction.kt))
- Write Robolectric-based tests verifying `appendText()` and `replaceText()` apply the correct text values
- Validate that the logic paths in the updated `updateText()` method execute correctly for both `isAppend=true` and `isAppend=false`

### Steps To Reproduce

1. Write a Robolectric test using `EditTextInputAction.appendText()` to input text into an `EditText`.
2. Set initial text in the `EditText` (e.g., "123").
3. Use `appendText("45")` and expect the final text to be "12345".
4. Run the test in Robolectric.
5. Observe that the text becomes "45" instead of "12345", indicating that `appendText()` is replacing rather than appending the text.

Concered file: [testing/src/main/java/org/oppia/android/testing/espresso/EditTextInputAction.kt](https://github.com/oppia/oppia-android/blob/d25fe1f0f29688f5ed485503bc0a9c4bfff2d73f/testing/src/main/java/org/oppia/android/testing/espresso/EditTextInputAction.kt#L22)

### Expected Behavior

When using `EditTextInputAction.appendText()` in Robolectric-based tests, the provided text should be appended to the existing contents of the `EditText`, preserving the original text.

Similarly, when using `replaceText()`, the original text in the `EditText` should be completely replaced with the new input.

These behaviors should match how Espresso handles `typeText()` and `replaceText()` on physical/emulated devices.

### Screenshots/Videos

_No response_

### What device/emulator are you using?

_No response_

### Which Android version is your device/emulator running?

_No response_

### Which version of the Oppia Android app are you using?

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by shrimpnaur**
This PR addresses part of #5818 by adding comprehensive tests for `EditTextInputAction` to verify correct append and replace text behavior in Robolectric environments.

## What This PR Does
- Adds `EditTextInputActionTest.kt` with comprehensive test coverage for `EditTextInputAction`
- Tests `appendText()` correctly appends to existing text (e.g., "123" + "45" = "12345")  
- Tests `replaceText()` correctly replaces existing text (e.g., "123" → "45")
- Covers edge cases like empty text, special characters, and mixed operations
- Ensures the fix from PR #5617 works correctly and prevents future regressions

## Testing
- All tests verify the behavior works correctly in Robolectric environments where the original bug occurred
- Tests cover both `isAppend=true` and `isAppend=false` code paths in the `updateText()` method
- Includes tests for sequential operations and edge cases

The core bug fix was already implemented in PR #5617. This PR adds the missing test coverage as specified in issue #5818.

Fixes part of #5818

## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

### CODE REVIEW

**Comment by shrimpnaur** [comment]
PTAL @theayushyadav11

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+    fun provideContext(application: Application): Context {
+      return application
+    }
+  }
+
+  @Singleton
+  @Component(
+    modules = [
+      TestModule::class, TestLogReportingModule::class, RobolectricModule::class,
+      TestDispatcherModule::class
```
Comment (theayushyadav11) [inline testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt]:
```suggestion
      TestModule::class, 
      TestLogReportingModule::class, 
      RobolectricModule::class,
      TestDispatcherModule::class
```

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    editText.setText("123")
+
+    editTextInputAction.replaceText("45").perform(null, editText)
+    testCoroutineDispatchers.runCurrent()
+
+    assertThat(editText.text.toString()).isEqualTo("45")
+  }
+
+  @Test
+  fun testReplaceText_multipleReplaces_replacesCorrectly() {
```
Comment (theayushyadav11) [inline testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt]:
```suggestion
  fun testReplaceText_onMultipleReplaces_replacesTextCorrectly() {
```

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    editTextInputAction.replaceText("First").perform(null, editText)
+    testCoroutineDispatchers.runCurrent()
+    editTextInputAction.replaceText("Second").perform(null, editText)
+    testCoroutineDispatchers.runCurrent()
+
+    assertThat(editText.text.toString()).isEqualTo("Second")
+  }
+
+  @Test
+  fun testAppendText_afterReplaceText_behavesCorrectly() {
```
Comment (theayushyadav11) [inline testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt]:
```suggestion
  fun testAppendText_afterReplaceText_appendsToReplacedText() {
```

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+    editTextInputAction.replaceText("New").perform(null, editText)
+    testCoroutineDispatchers.runCurrent()
+    editTextInputAction.appendText("Text").perform(null, editText)
+    testCoroutineDispatchers.runCurrent()
+
+    assertThat(editText.text.toString()).isEqualTo("NewText")
+  }
+
+  @Test
+  fun testReplaceText_afterAppendText_behavesCorrectly() {
```
Comment (theayushyadav11) [inline testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt]:
```suggestion
  fun testReplaceText_afterAppendText_replacesEntireText() {
```

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+    editText.setText("")
+
+    editTextInputAction.replaceText("Hello").perform(null, editText)
+    testCoroutineDispatchers.runCurrent()
+
+    assertThat(editText.text.toString()).isEqualTo("Hello")
+  }
+
+  @Test
+  fun testReplaceText_withExistingText_replacesCorrectly() {
```
Comment (theayushyadav11) [inline testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt]:
```suggestion
  fun testReplaceText_withExistingText_replacesTextCorrectly() {
```

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+    editText.setText("123")
+
+    editTextInputAction.appendText("45").perform(null, editText)
+    testCoroutineDispatchers.runCurrent()
+
+    assertThat(editText.text.toString()).isEqualTo("12345")
+  }
+
+  @Test
+  fun testAppendText_multipleAppends_appendsCorrectly() {
```
Comment (theayushyadav11) [inline testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt]:
```suggestion
  fun testAppendText_onMultipleAppends_appendsCorrectly() {
```

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+    editText.setText("")
+
+    editTextInputAction.appendText("Hello").perform(null, editText)
+    testCoroutineDispatchers.runCurrent()
+
+    assertThat(editText.text.toString()).isEqualTo("Hello")
+  }
+
+  @Test
+  fun testAppendText_withExistingText_appendsCorrectly() {
```
Comment (theayushyadav11) [inline testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt]:
```suggestion
  fun testAppendText_withExistingText_appendsTextCorrectly() {
```

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+    editTextInputAction.appendText("B").perform(null, editText)
+    testCoroutineDispatchers.runCurrent()
+    editTextInputAction.appendText("C").perform(null, editText)
+    testCoroutineDispatchers.runCurrent()
+
+    assertThat(editText.text.toString()).isEqualTo("ABC")
+  }
+
+  @Test
+  fun testAppendText_withSpaces_appendsCorrectly() {
```
Comment (theayushyadav11) [inline testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt]:
```suggestion
  fun testAppendText_withSpaces_appendsTextCorrectly() {
```

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+    editTextInputAction.appendText("Added").perform(null, editText)
+    testCoroutineDispatchers.runCurrent()
+    editTextInputAction.replaceText("Replaced").perform(null, editText)
+    testCoroutineDispatchers.runCurrent()
+
+    assertThat(editText.text.toString()).isEqualTo("Replaced")
+  }
+
+  @Test
+  fun testAppendText_withSpecialCharacters_appendsCorrectly() {
```
Comment (theayushyadav11) [inline testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt]:
```suggestion
  fun testAppendText_withSpecialCharacters_appendsTextCorrectly() {
```

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/**
+ * Tests for [EditTextInputAction] to verify correct append and replace text behavior
+ * in Robolectric environments.
+ */
+@RunWith(AndroidJUnit4::class)
+@Config(application = EditTextInputActionTest.TestApplication::class)
+class EditTextInputActionTest {
```
Comment (theayushyadav11) [inline testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt:30]: Nit: To maintain consistency, please arrange the tests in the following order:

1.All `appendText()` tests.
2.All `replaceText()` tests.

**Review by theayushyadav11 — CHANGES_REQUESTED** [review]
Thanks @shrimpnaur, this is nicely done. Just a few nits regarding test namings ([guide](https://github.com/oppia/oppia-android/wiki/Oppia-Android-Testing#naming-convention)) and arrangement.

**Comment by theayushyadav11** [comment]
Hi @shrimpnaur, just a small request, please avoid resolving comments that were not opened by you. Instead, you can respond with "Done," "Updated," or share any thoughts you have. The reviewer will take care of resolving the comments.
Thanks

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt (hunk 1 above)
Comment (shrimpnaur) [reply testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt]: Done.

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt (hunk 2 above)
Comment (shrimpnaur) [reply testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt]: Done.

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt (hunk 3 above)
Comment (shrimpnaur) [reply testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt]: Done.

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt (hunk 4 above)
Comment (shrimpnaur) [reply testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt]: Done.

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt (hunk 5 above)
Comment (shrimpnaur) [reply testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt]: Done.

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt (hunk 6 above)
Comment (shrimpnaur) [reply testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt]: Done.

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt (hunk 7 above)
Comment (shrimpnaur) [reply testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt]: Done

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt (hunk 8 above)
Comment (shrimpnaur) [reply testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt]: Done.

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt (hunk 9 above)
Comment (shrimpnaur) [reply testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt]: Done.

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt (hunk 10 above)
Comment (shrimpnaur) [reply testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt:30]: Done.

**Comment by shrimpnaur** [comment]
Understood and resolved the comments. PTAL @theayushyadav11

**Comment by shrimpnaur** [comment]
@theayushyadav11 Hey there, just wanted to check up on this. Does everything work as intended?

**Comment by adhiamboperes** [comment]
Hi @shrimpnaur, apologies for the delay. We will give you a full review tomorrow.

**Comment by theayushyadav11** [comment]
Thanks @shrimpnaur for addressing the comments. This LGTM.
I have updated the branch to adhere to latest `develop` branch.
@adhiamboperes PTAL for the final pass.

**Review by theayushyadav11 — APPROVED** [review]

File: testing/src/test/java/org/oppia/android/testing/espresso/BUILD.bazel
Comment (adhiamboperes) [inline testing/src/test/java/org/oppia/android/testing/espresso/BUILD.bazel:1]: This file is not needed. The project already indexes the targets in this package in https://github.com/oppia/oppia-android/blob/b11e472eeb852e155b2f7720b9e39e77871cdafc/app/BUILD.bazel#L753.

We are doing ongoing modularization work to ensure each package will have its own BUILD file like you created, but it is not finished yet.

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+import org.robolectric.annotation.Config
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/**
+ * Tests for [EditTextInputAction] to verify correct append and replace text behavior
+ * in Robolectric environments.
+ */
+@RunWith(AndroidJUnit4::class)
+@Config(application = EditTextInputActionTest.TestApplication::class)
```
Comment (adhiamboperes) [inline testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt:29]: These tests are good if the goal is lightweight unit tests, but the common pattern in the repo is to use instrumented unit tests/ semi integration tests like in [TextInputActionTest.kt](https://github.com/oppia/oppia-android/blob/b11e472eeb852e155b2f7720b9e39e77871cdafc/testing/src/test/java/org/oppia/android/testing/espresso/TextInputActionTest.kt).

I would like @BenHenning to weigh in here, whether we should go with this approach.

**Review by adhiamboperes — CHANGES_REQUESTED** [review]
Thanks @shrimpnaur for this PR. I am handing this over to @BenHenning to help with the technical decisions.

**Comment by shrimpnaur** [comment]
@BenHenning @adhiamboperes   
  
Thank you for the feedback. I understand that these are architectural decisions that need your input.   
  
Could you please let me know if I should wait for the technical decisions before making any changes?  
and also if there is there anything I can help clarify about my current implementation while you're reviewing the approach?  
 
I'm happy to adjust the implementation once the preferred approach is determined.

**Comment by adhiamboperes** [comment]
Hi @shrimpnaur, I understand that you are blocked on this PR. It would be great to work on a new issue in the meantime, to parallelize your efforts. We have updated our [list of starter issues](https://github.com/oppia/oppia-android/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22good%20first%20issue%22%20no%3Aassignee), so I hope you find some interesting stuff to work on!

**Comment by shrimpnaur** [comment]
@adhiamboperes Yeah sure. :+1:

File: testing/src/test/java/org/oppia/android/testing/espresso/BUILD.bazel
Comment (BenHenning) [reply testing/src/test/java/org/oppia/android/testing/espresso/BUILD.bazel:1]: This is probably fine to include though. It seems correct from what I can tell.

File: testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt (hunk 11 above)
Comment (BenHenning) [reply testing/src/test/java/org/oppia/android/testing/espresso/EditTextInputActionTest.kt:29]: I agree. The variety of tests is really good, but we ought to ensure that the action actually integrates with Espresso correctly (which means using it as an Espresso action rather than directly calling it). `TextInputActionTest` is a really good basis for reference on how to do that since it's also validating a custom Espresso action.

**Review by BenHenning — COMMENTED** [review]
Thanks @shrimpnaur! Apologies for the delay--I was sick and it took a number of days to catch up on things. I've followed up to @adhiamboperes's comments--PTAL.

**Comment by shrimpnaur** [comment]
Hi @BenHenning ,
Thanks for the clarification! I’ll leave the BUILD.bazel file as is and update the my test by following the pattern in TextInputActionTest.kt so that there is proper integration with Espresso. 

I'll make these changes and update the PR shortly.  
Thanks for the guidance!

**Comment by adhiamboperes** [comment]
Hi @shrimpnaur, do you need support on this PR?

**Comment by shrimpnaur** [comment]
Hey @adhiamboperes 
I'm working on updating the tests, but I need guidance on which test activity to use. Should I:  
1. Use an existing test activity like InputInteractionViewTestActivity?  
2. Create a new dedicated test activity for EditTextInputAction?  
  
Could you point me to the best approach for this?  
  
Thanks!

**Comment by adhiamboperes** [comment]
Hi @shrimpnaur, please create a new test activity as it is fairly simple. Also make sure you add the activity to the `test_file_exemptions` file, as well as to the testing module's `AndroidManifest` and the CODEOWNERS file, following the example of `TextInputActionTestActivity`.

**Comment by adhiamboperes** [comment]
@shrimpnaur, will you be reopening this PR?
